File: app/src/main/java/com/aashutosh/desimall_pro/utils/OTP_Receiver.java
package com.aashutosh.desimall_pro.utils;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import com.google.android.gms.auth.api.phone.SmsRetriever;
import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.common.api.Status;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class OTP_Receiver extends BroadcastReceiver {

    private OTPReceiverListener otpReceiverListener;

    public OTP_Receiver() {
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (SmsRetriever.SMS_RETRIEVED_ACTION.equals(intent.getAction())) {

            Bundle bundle = intent.getExtras();
            if (bundle != null) {
                Status status = (Status) bundle.get(SmsRetriever.EXTRA_STATUS);
                if (status != null) {
                    switch (status.getStatusCode()) {
                        case CommonStatusCodes.SUCCESS:
                            String message = (String) bundle.get(SmsRetriever.EXTRA_SMS_MESSAGE);
                            if (message != null) {
                                Pattern pattern = Pattern.compile("\\d{6}");
                                Matcher matcher = pattern.matcher(message);
                                if (matcher.find()) {
                                    String myOtp = matcher.group(0);
                                    if (this.otpReceiverListener != null) {
                                        this.otpReceiverListener.onOtpSuccess(myOtp);
                                    } else {
                                        if (this.otpReceiverListener != null) {
                                            this.otpReceiverListener.onOtpTimeOut();
                                        }
                                    }
                                }
                            }
                            break;
                        case CommonStatusCodes.TIMEOUT:
                            if (this.otpReceiverListener != null) {
                                this.otpReceiverListener.onOtpTimeOut();
                            }
                            break;


                    }
                }
            }
        }
    }

    public interface OTPReceiverListener {
        void onOtpSuccess(String otp);

        void onOtpTimeOut();

    }
}

File: app/src/main/java/com/aashutosh/desimall_pro/utils/OTPReceiver.kt
package com.aashutosh.desimall_pro.utils

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import com.aashutosh.desimall_pro.utils.PushNotificationService.TAG
import com.google.android.gms.auth.api.phone.SmsRetriever
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.api.Status
import java.util.regex.Pattern
import kotlin.math.log

class OTPReceiver : BroadcastReceiver() {

    private var otpReceiverListener: OTPReceiverListener? = null

    fun initListener(oTPReceiverListener: OTPReceiverListener) {
        otpReceiverListener = oTPReceiverListener
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (SmsRetriever.SMS_RETRIEVED_ACTION == intent.action) {
            val extras = intent.extras
            if (extras != null) {
                val status = extras.getSerializable(SmsRetriever.EXTRA_STATUS) as? Status
                    ?: extras.getParcelable(SmsRetriever.EXTRA_STATUS)

                if (status != null) {
                    when (status.statusCode) {
                        CommonStatusCodes.SUCCESS -> {
                            val message = extras.getString(SmsRetriever.EXTRA_SMS_MESSAGE)
                            if (!message.isNullOrEmpty()) {
                                val pattern = Pattern.compile("\\b\\d{6}\\b")
                                val matcher = pattern.matcher(message)
                                if (matcher.find()) {
                                    val myOtp = matcher.group(0)
                                    otpReceiverListener?.onOtpSuccess(myOtp)
                                    Log.d(TAG, "fuji: it should work ")
                                }else{
                                    otpReceiverListener?.onOtpTimeOut()
                                    Log.d(TAG, "fuji: matcher problem ")
                                }
                            }else{
                                Log.d(TAG, "fuji: message is empty or null ")
                                otpReceiverListener?.onOtpTimeOut()
                            }
                        }

                        CommonStatusCodes.TIMEOUT -> otpReceiverListener?.onOtpTimeOut()
                    }
                }
            }
        }
    }

    interface OTPReceiverListener {
        fun onOtpSuccess(otp: String?)
        fun onOtpTimeOut()
    }
}
